x-pack/qa/full-cluster-restart/src/javaRestTest/java/org/elasticsearch/xpack/restart/AbstractXpackFullClusterRestartTestCase.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.restart;
+
+import org.elasticsearch.test.cluster.ElasticsearchCluster;
+import org.elasticsearch.test.cluster.FeatureFlag;
+import org.elasticsearch.test.cluster.local.distribution.DistributionType;
+import org.elasticsearch.test.cluster.util.resource.Resource;
+import org.elasticsearch.upgrades.FullClustRestartUpgradeStatus;
+import org.elasticsearch.upgrades.ParameterizedFullClusterRestartTestCase;
+import org.junit.ClassRule;
+
+public abstract class AbstractXpackFullClusterRestartTestCase extends ParameterizedFullClusterRestartTestCase {
+
+    @ClassRule
+    public static ElasticsearchCluster cluster = ElasticsearchCluster.local()
```

Comment:
The duplication across test classes is a little alarming. We discussed this before, but are there any plans/ideas for how we can better compose these test setups? The security setup in particular is duplicated *a lot* already across gradle code for testclusters, and it seems to be just as bad, if not worse, due to the static configuration that cannot be inherited.

Reply:
I'm not sure what you mean. This is an abstract class and all the other test cases extend from it so there's no duplication here.

Reply:
Also, there are ways to allow child classes to add or override config. This is exactly what's done in the `CoreFullClusterRestartIT` that uses the cluster from the non-xpack project but then just adds security.

https://github.com/elastic/elasticsearch/blob/a14906c42e4d05e22e0dbdfd55d51d206f97d4dc/x-pack/qa/full-cluster-restart/src/javaRestTest/java/org/elasticsearch/xpack/restart/CoreFullClusterRestartIT.java#L23-L33

Reply:
Configuring security is probably the most tedious bit. It's done explicitly in every project. We can _definitely_ consolidate this and @ywangd has brought this up as well. There's no need to have a billion different certs and keystores all thorughout our codebase for testing. We could just use a single one, and then just configure a `SettingsProvider` on the cluster that encapsulated all that config.

Reply:
Some tests intentionally need different certs or same cert in different formats. But l think many of them can indeed share these files. So something like:
* Separate key/cert/ca for the Transport, Http and the new RemoteCluster interfaces
* Above key/cert/ca in serveral different formats PEM (needed for FIPS), PKCS12, JKS. 

This should help reduce the configuration ceremony needed for security. But it may not be enough to reduce the sheer number of key/cert/ca files. Because those tests that cannot share these files still require a large number of them to be pre-generated. Probably another possibility is to generate them on the fly. Since these are REST tests, I think we should have access to the `elasticsearch-certutil` command which is used to generate many of the key/cert/ca. This plus Java keytool and zip file manipulation could provide us quite a bit to play with.